### PR TITLE
Issue#256 extend user creation with invite token

### PR DESCRIPTION
**Pull request by IRoss-git**
dev
## ZenHub

* [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009/issues/ita-social-projects/home/256)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -29,42 +33,74 @@
...
+
+        var invitation = invitationRepository.findInvitationByRegistrationToken(userDto.getRegistrationToken())
+                .filter(invitation1 -> invitation1.getStatus().equals(InvitationStatus.PROCESSING))
+                .orElseThrow(() -> new NotFoundHomeException("Invitation with provided token not found"));
+
+        validateEmailsMatching(invitation.getEmail(), userDto.getEmail());
+
+        if (invitation.getType().equals(InvitationType.APARTMENT)) {
+            apartmentInvitationService.acceptUserInvitation(invitation);
+        } else {
```
Comment (Illia-Shchur) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java]: Please add check for another invitation status. If we may add another one in future, this is a potential place for a bug.

**Review by Illia-Shchur — APPROVED** [review]

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java (hunk 1 above)
Comment (IRoss-git) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/UserServiceImpl.java]: Thanks for notice. Added switch statement

**Review by Vladyslav-Frolov — APPROVED** [review]

**Comment by Vladyslav-Frolov** [comment]
> dev
> 
> ## ZenHub
> * [Main ZenHub ticket](https://app.zenhub.com/workspaces/home-project-5f7b77ff8db73a001cb17009)
> 
> ## Code reviewers
> * [ ]  @github_username
> 
> ### Second Level Review
> * [ ]  @github_username
> 
> ## Summary of issue
> ToDo
> 
> ## Summary of change
> ToDo
> 
> ## Testing approach
> ToDo
> 
> ## CHECK LIST
> * [ ]   СI passed
> * [ ]   Сode coverage >=95%
> * [ ]   PR is reviewed manually again (to make sure you have 100% ready code)
> * [ ]   All reviewers agreed to merge the PR
> * [ ]   I've checked new feature as logged in and logged out user if needed
> * [ ]   PR meets all conventions

No description has been added to the pull request, just a template → must be filled in.

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/ownerships/OwnershipApiIT.java [hunk 2]
```diff
@@ -2,14 +2,26 @@
 
 import com.softserveinc.ita.homeproject.ApiException;
 import com.softserveinc.ita.homeproject.ApiResponse;
-import com.softserveinc.ita.homeproject.api.ApartmentOwnershipApi;
+import com.softserveinc.ita.homeproject.api.*;
```
Comment (likeRewca) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/ownerships/OwnershipApiIT.java]: even if this module does not check the checklist, let's adhere to the general standards

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/users/QueryUserIT.java [hunk 3]
```diff
@@ -5,27 +5,38 @@
...
 import com.softserveinc.ita.homeproject.api.UserApi;
 import com.softserveinc.ita.homeproject.api.tests.query.UserQuery;
 import com.softserveinc.ita.homeproject.api.tests.utils.ApiClientUtil;
 import com.softserveinc.ita.homeproject.api.tests.utils.QueryFilterUtils;
-import com.softserveinc.ita.homeproject.model.CreateUser;
-import com.softserveinc.ita.homeproject.model.ReadUser;
+import com.softserveinc.ita.homeproject.api.tests.utils.mail.mock.ApiMailHogUtil;
+import com.softserveinc.ita.homeproject.api.tests.utils.mail.mock.ApiUsageFacade;
+import com.softserveinc.ita.homeproject.api.tests.utils.mail.mock.dto.MailHogApiResponse;
+import com.softserveinc.ita.homeproject.model.*;
```
Comment (likeRewca) [inline home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/users/QueryUserIT.java]: same here

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/impl/HomeUserDetailsService.java [hunk 4]
```diff
@@ -39,13 +40,21 @@ public HomeUserDetailsService(UserRepository userRepository) {
...
     public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
-        User user = userRepository.findByEmail(email)
-            .orElseThrow(() -> new UsernameNotFoundException("There is no user with given email!"));
-        Set<Permission> perms = user.getRoles().stream()
-            .map(Role::getPermissions)
-            .flatMap(Collection::stream)
-            .collect(Collectors.toSet());
-
-        return new HomeUserWrapperDetails(user, perms);
+        var user = userRepository.findByEmail(email)
```
Comment (likeRewca) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/impl/HomeUserDetailsService.java:44]: what about enabled user?

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/ownerships/OwnershipApiIT.java (hunk 2 above)
Comment (IRoss-git) [reply home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/ownerships/OwnershipApiIT.java]: done

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/users/QueryUserIT.java (hunk 3 above)
Comment (IRoss-git) [reply home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/users/QueryUserIT.java]: done

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/impl/HomeUserDetailsService.java (hunk 4 above)
Comment (IRoss-git) [reply home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/impl/HomeUserDetailsService.java:44]: done

File: home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/users/QueryUserIT.java (hunk 3 above)
Comment (likeRewca) [reply home-api-tests/src/test/java/com/softserveinc/ita/homeproject/api/tests/users/QueryUserIT.java]: here u are not change

File: home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/RoleEnum.java [hunk 5]
```diff
@@ -9,7 +9,9 @@
 
     ADMIN("admin"),
     USER("user"),
```
Comment (likeRewca) [inline home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/RoleEnum.java:11]: As Oleksandr said that we can delete role User

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/dto/RoleDto.java [hunk 6]
```diff
@@ -4,7 +4,8 @@
 
     ADMIN("admin"),
     USER("user"),
```
Comment (likeRewca) [inline home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/dto/RoleDto.java:6]: delete here too

File: home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/RoleEnum.java (hunk 5 above)
Comment (IRoss-git) [reply home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/RoleEnum.java:11]: It should be deleted in role matrix impl

File: home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/dto/RoleDto.java (hunk 6 above)
Comment (IRoss-git) [reply home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/dto/RoleDto.java:6]: It should be deleted in role matrix impl

**Review by likeRewca — APPROVED** [review]

**Review by AriannaLiss — APPROVED** [review]

File: home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/HomeUserWrapperDetails.java [hunk 7]
```diff
@@ -31,8 +30,8 @@ public HomeUserWrapperDetails(User user, Collection<Permission> userPermissions)
     public Collection<? extends GrantedAuthority> getAuthorities() {
         Set<GrantedAuthority> authorities;
         authorities = userPermissions.stream()
-            .map(permission -> new SimpleGrantedAuthority(permission.getName()))
-            .collect(Collectors.toSet());
+                .map(SimpleGrantedAuthority::new)
```
Comment (IhorSamoshost) [inline home-application/src/main/java/com/softserveinc/ita/homeproject/application/config/HomeUserWrapperDetails.java:33]: Will this mapping assign a role value to the field "String role" in the new SimpleGrantedAuthority object?

File: home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/UserCooperation.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    @JoinColumn(name = "user_id")
+    private User user;
+
+    @ManyToOne
+    @JoinColumn(name = "cooperation_id")
+    private Cooperation cooperation;
+
+    @ManyToOne
+    @JoinColumn(name = "role_id")
+    private Role role;
```
Comment (kh0ma) [inline home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/UserCooperation.java:29]: I think it should be list as we will be able to have more than one role per user in one cooperation.

File: home-open-api/src/main/resources/yaml/openapi.yaml [hunk 9]
```diff
@@ -1875,6 +1880,12 @@ components:
           example: "77778"
           minLength: 1
           maxLength: 50
+        admin_email:
```
Comment (kh0ma) [inline home-open-api/src/main/resources/yaml/openapi.yaml]: Why do we need it? I think this information must be hidden until admin creates a contact and expose the email address explicitly. See `contacts` property.

File: home-service/pom.xml [hunk 10]
```diff
@@ -50,6 +50,11 @@
             <artifactId>validation-api</artifactId>
             <version>2.0.1.Final</version>
         </dependency>
+        <dependency>
+            <groupId>com.fasterxml.uuid</groupId>
+            <artifactId>java-uuid-generator</artifactId>
+            <version>4.0.1</version>
```
Comment (kh0ma) [inline home-service/pom.xml]: Please move version to properties section.

**Review by kh0ma — CHANGES_REQUESTED** [review]

File: home-service/pom.xml (hunk 10 above)
Comment (IRoss-git) [reply home-service/pom.xml]: Moved

File: home-open-api/src/main/resources/yaml/openapi.yaml (hunk 9 above)
Comment (IRoss-git) [reply home-open-api/src/main/resources/yaml/openapi.yaml]: Thanks, forgot to delete

**Review by kizimov — APPROVED** [review]

**Review by Vladyslav-Frolov — APPROVED** [review]

File: home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/UserCooperation.java (hunk 8 above)
Comment (IRoss-git) [reply home-data/src/main/java/com/softserveinc/ita/homeproject/homedata/entity/UserCooperation.java:29]: I checked and it working as we discussed thus user in cooperation can have more than one role

**Review by kh0ma — APPROVED** [review]

**Pull request merged**